src/sources/stdin.rs

```diff
@@ -82,102 +70,40 @@ pub fn stdin_source<R>(
 where
     R: Send + io::BufRead + 'static,
 {
-    // The idea is to have one dedicated future for reading stdin running in the background,
-    // and the sources would recieve the lines thorugh a multi consumer channel.
-    //
-    // Implemented solution relies on having a copy of optional sender behind a global mutex,
-    // and have stdin sources and background thread synchronize on it.
-    //
-    // When source is built it must:
-    // 1. enter critical section by locking mutex.
-    // 2. if sender isn't present start background thread and set sender.
-    // 3. gain receiver by calling subscribe on the sender.
-    // 4. release lock.
-    //
-    // When source has finished it should just drop it's receiver.
-    //
-    // Background thread should be started with copy of the sender.
-    //
-    // Once background thread wants to stop it must:
-    // 1. enter critical section by locking mutex.
-    // 2. if there are receivers it must abort the stop.
-    // 3. remove sender.
-    // 4. release lock.
-    //
-    // Although it's possible to implement this in a lock free, maybe even wait free manner,
-    // this should be easier to reason about and performance shouldn't suffer since this procedure
-    // is cold compared to the rest of the source.
-
     let host_key = config
         .host_key
         .unwrap_or_else(|| event::log_schema().host_key().to_string());
     let hostname = hostname::get_hostname();
 
-    let mut guard = CRITICAL_SECTION
-        .lock()
-        .map_err(|_| BuildError::CriticalSectionPoisoned)?;
-    let receiver = match guard.as_ref() {
-        Some(sender) => sender.subscribe(),
-        None => {
-            let (sender, receiver) = channel(1024);
-            *guard = Some(sender.clone());
-
-            // Start the background thread
-            thread::spawn(move || {
-                info!("Capturing STDIN.");
+    let (mut sender, receiver) = channel(1024);
 
-                for line in stdin.lines() {
-                    match line {
-                        Err(error) => {
-                            emit!(StdinReadFailed { error });
-                            break;
-                        }
-                        Ok(line) => {
-                            if sender.send(Bytes::from(line)).is_err() {
-                                // There are no active receivers.
-                                // Try to stop.
-                                let mut guard =
-                                    CRITICAL_SECTION.lock().expect("CRITICAL_SECTION poisoned");
+    // Start the background thread
+    thread::spawn(move || {
```

Comment:
It's not related with this PR, but should we use [tokio::task::spawn_blocking](https://docs.rs/tokio/0.2.22/tokio/task/fn.spawn_blocking.html) everywhere instead `std::thread`? Currently, we use both options.

Reply:
It's a good question, I think it makes sense to use a thread for a case like this, where we have a loop inside. From the docs:

> This function is intended for non-async operations that eventually finish on their own. If you want to spawn an ordinary thread, you should use `thread::spawn` instead.

In this case the operation does finish on it's own though. I think it all depends on if it's "ok" to access the executor from the inside of `tokio::task::spawn_blocking`.

Either way, it's not clear to me either, so I'm looking forward to figuring this out.

Reply:
My understanding is that `spawn_blocking` is more for situations where a relatively fixed amount of work is done before the task completes. In our case, we are spawning a loop that runs essentially for the life of the process so I think a normal thread is more appropriate.